### DISCUSSION TITLE
How to filter (select) on stream provider ?

### DISCUSSION THREAD

**Discussion by pbouttier** [rrousselGit/riverpod#1989] [Q&A] [answered]
Hi,

I have a User (class) with different parameters (age, name, ...). I have different buttons to modify this or that property of this user. How to filter the values to listen and limit the number of rebuilds?

For example : 

```
final userSettingsProvider = StreamProvider.autoDispose<UserSettingsModel>((ref) {
  return ref.watch(databaseProvider).userSettingsStream();
});
```

```
DropdownButton<Gender>(
  value: ref.watch(userSettingsProvider.select((stream) => stream?.value?.gender)),
  onChanged: (Gender newValue) {
    //...
    ref.read(databaseProvider).updateUserSettings(user);
  },
  items: <Gender>[Gender.man, Gender.woman].map((Gender value) {
    return DropdownMenuItem<Gender>(
      value: value,
      child: Text(value.toString()),
    );
  }).toList(),
),
DropdownButton<Unit>(
  value: ref.watch(userSettingsProvider.select((stream) => stream?.value?.defaultUnit)),
  onChanged: (Unit newValue) {
    //...
   ref.read(databaseProvider).updateUserSettings(user);
  },
  items: <Unit>[Unit.metric, Unit.angloSaxon].map((Unit value) {
    return DropdownMenuItem<Unit>(
      value: value,
      child: Text(value.toString()),
    );
  }).toList(),
)
```

In this case i try to filter on a single value at time, but when a update A or B, both dropdown are rebuilds. The DB is firebase.

Is it that select can only filter at one level, or that the stream.value is recreated and therefore we can't filter on the final value?
Maybe make a stateprovider instead of watch the streamProvider (`userSettingsProvider`) value ?
Maybe go for a streambuilder ?
Maybe i miss a important concept ?

Thanks for your help

PS : can be a duplicate of #1083 ?

**Comment by rrousselGit** [comment, marked as answer]
Maybe you want to wrap the dropdown in a `Consumer`?

```dart
Consumer(builder: (context, ref, _) {
  return Dropdown(...)
});
```

This will allow your dropdowns to rebuild independently from each other.

**Comment by pbouttier** [comment]
Yes it works thanks.

But it's a lot of boilerplates to add a consumer above each widget, isn't it possible with a simple select ?

**Reply by rrousselGit** [reply]
No. The way you wrote your widget, they will always rebuild together. It's not really Riverpod the issue

They need to be in a different "build" method, which is what that Consumer does. Although you could alternatively extract the dropdown code in a separate widget

**Comment by pbouttier** [comment]
Ok i see.

Thanks a lot for your answer and your repo.
